MM-14104 Update logic for Markdown emphasis from upstream

hmhealey opened the pull request:
For context, we don't have our marked fork up to date with upstream because we've made some extensive changes to it internally. Since the goal is still to switch to commonmark.js, there haven't been many efforts to get it up to date, but occasionally we need to manually copy changes like this to fix annoying bugs.

The actual changes for this PR are here: https://github.com/mattermost/marked/commit/c1b6989f5cbacba6dceaabe8eabd7be8ec655cd6

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-14104

deanwhillier commented:
Looks like the update is causing some tests to fail. @hmhealey is investigating.

hmhealey commented:
Fixed. Turns out we were previously using a branch on the marked fork, and switching to the master branch on it lost some changes. I merged those into master and updated the commit.

Also, one of the tests was legitimately broken
